platform/darwin/test/MGLPredicateTests.mm

```diff
@@ -23,558 +23,404 @@ @interface MGLPredicateTests : XCTestCase
+        XCTAssertEqualObjects(predicate.mgl_jsonExpressionObject, jsonExpression);
+        [self testSymmetryWithPredicate:predicate
+                          mustRoundTrip:NO];
     }
-    
     {
-        auto actual = [NSPredicate predicateWithFormat:@"%@ CONTAINS %K", @[@"LineString", @"Polygon"], @"$type"].mgl_filter;
-        mbgl::style::TypeInFilter expected = { .values = { mbgl::FeatureType::LineString, mbgl::FeatureType::Polygon } };
-        MGLAssertEqualFilters(actual, expected);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"$mgl_featureIdentifier != 67086180"];
+        NSArray *jsonExpression = @[@"!=", @[@"id"], @67086180];
+        XCTAssertEqualObjects(predicate.mgl_jsonExpressionObject, jsonExpression);
+        [self testSymmetryWithPredicate:predicate
+                          mustRoundTrip:NO];
     }
-    
     {
-        auto actual = [NSPredicate predicateWithFormat:@"{67086180, 3709678893, 3352016856, 4189833989} CONTAINS %K", @"$id"].mgl_filter;
-        mbgl::style::IdentifierInFilter expected = { .values = { UINT64_C(67086180), UINT64_C(3709678893), UINT64_C(3352016856), UINT64_C(4189833989) } };
-        MGLAssertEqualFilters(actual, expected);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"$mgl_featureIdentifier != nil"];
+        NSArray *jsonExpression = @[@"!=", @[@"id"],  [NSNull null]];
+        XCTAssertEqualObjects(predicate.mgl_jsonExpressionObject, jsonExpression);
+        [self testSymmetryWithPredicate:predicate
+                          mustRoundTrip:NO];
     }
-    
     {
-        auto actual = [NSPredicate predicateWithFormat:@"%@ CONTAINS %K", @[@67086180, @3709678893, @3352016856, @4189833989], @"$id"].mgl_filter;
-        mbgl::style::IdentifierInFilter expected = { .values = { UINT64_C(67086180), UINT64_C(3709678893), UINT64_C(3352016856), UINT64_C(4189833989) } };
-        MGLAssertEqualFilters(actual, expected);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"a != 'b'"];
+        NSArray *jsonExpression = @[@"!=", @[@"get", @"a"], @"b"];
+        XCTAssertEqualObjects(predicate.mgl_jsonExpressionObject, jsonExpression);
+        [self testSymmetryWithPredicate:predicate
+                          mustRoundTrip:NO];
     }
-
-    XCTAssertThrowsSpecificNamed([NSPredicate predicateWithFormat:@"a CONTAINS 'Mapbox'"].mgl_filter, NSException, NSInvalidArgumentException);
-
     {
-        auto actual = [NSPredicate predicateWithFormat:@"a == 'b' AND c == 'd'"].mgl_filter;
-        mbgl::style::AllFilter expected = {
-            .filters = {
-                mbgl::style::EqualsFilter { .key = "a", .value = std::string("b") },
-                mbgl::style::EqualsFilter { .key = "c", .value = std::string("d") },
-            },
-        };
-        MGLAssertEqualFilters(actual, expected);
+        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"a != nil"];
```

Comment:
This used to be converted into a `NotHasFilter`. Ideally, we’d convert it to `["!", ["has", ["get", "a"]]]`. Are we sure that `["!-", ["get", "a"], null]` has the desired effect?

Reply:
`["==", ["get", "a"], null]` should be equivalent to `["!", ["has", "a"]]`, because `get` returns `null` for a missing key

Reply:
That said, the latter is probably a more direct translation..